bolt-contracts/src/contracts/BoltChallenger.sol

```diff
@@ -0,0 +1,357 @@
+    }
+
+    /// @notice Resolve a challenge by providing proofs of the inclusion of the committed transaction.
+    /// @dev Challenges are DEFENDED if the resolver successfully defends the inclusion of the transaction,
+    /// and LOST if the challenger successfully demonstrates that the inclusion commitment was breached or
+    /// enough time has passed without proper resolution.
+    ///
+    /// q: should we also have a commit-reveal scheme for resolutions to avoid frontrunning to steal bonds?
+    function _resolve(bytes32 challengeID, bytes32 trustedBlockHash, Proof calldata proof) internal {
+        // The challenge is assumed to exist at this point, so we can safely access it.
+        Challenge storage challenge = challenges[challengeID];
+
+        if (challenge.status != ChallengeStatus.Open) {
+            revert ChallengeAlreadyResolved();
+        }
+
+        if (challenge.openedAt + MAX_CHALLENGE_DURATION < Time.timestamp()) {
+            // If the challenge has expired without being resolved, it is considered lost.
+            challenge.status = ChallengeStatus.Lost;
+            _transferFullBond(challenge.challenger);
+            emit ChallengeLost(challengeID);
+            return;
+        }
+
+        // Verify the validity of the header against the trusted block hash.
+        if (keccak256(proof.blockHeaderRLP) != trustedBlockHash) {
+            revert InvalidBlockHash();
+        }
+
+        // Decode the RLP-encoded block header fields
+        BlockHeaderData memory blockHeader = _decodeBlockHeaderRLP(proof.blockHeaderRLP);
+
+        // Decode the committed raw signed transaction. Its sender will be the account to prove existence of.
+        TransactionDecoder.Transaction memory decodedTx = challenge.commitment.signedTx.decodeEnveloped();
+
+        // Decode the account fields by checking the account proof against the state root of the block header.
+        // The key in the account trie is the account pubkey (address) which we can recover from the signed tx.
+        (bool accountExists, bytes memory accountRLP) = SecureMerkleTrie.get(
+            abi.encodePacked(decodedTx.recoverSender()), proof.accountMerkleProof, blockHeader.stateRoot
+        );
+
+        if (!accountExists) {
+            revert AccountDoesNotExist();
+        }
+
+        AccountData memory account = _decodeAccountRLP(accountRLP);
+
+        if (account.nonce > decodedTx.nonce) {
+            // The transaction recovered sender has sent a transaction with a higher nonce than the committed
+            // transaction, before the proposer could include it. Consider the challenge defended, as the
+            // proposer is not at fault. The bond will be shared between the resolver and commitment signer.
+            challenge.status = ChallengeStatus.Defended;
+            _transferHalfBond(msg.sender);
+            _transferHalfBond(challenge.target);
+            emit ChallengeDefended(challengeID);
```

Comment:
Transferring part of the bond to an entity different than the commitment signer (which is the user who requested the preconf) might incentive frontrunning the user, doesn't it?
I think the user has already all the incentives to do it by itself.

This is kinda an open question so I'd drop a TODO comment above this logic

Reply:
> I think the user has already all the incentives to do it by itself.

it does incentivise front-running, hence the question comment on the _resolve 
function if we should use a commit and reveal scheme. 

This incentive makes sense in my mind because we want a healthy and speedy processing of challenges.
From a mechanism perspective it's better to give a reward to the resolver (even if less than 0.5 ETH) IMO

Reply:
I'll solve the comment for now and we can discuss this further privately